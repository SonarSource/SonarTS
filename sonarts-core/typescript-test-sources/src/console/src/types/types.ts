import {Step} from './gettingStarted'
import {TypedValue} from './utils'

export interface RelayConnection<T> {
  edges: RelayEdge<T>[]
}

interface RelayEdge<T> {
  node: T
}

export interface Viewer {
  id: string
  user: Customer
  model: Model
  project: Project
  crm: any
}

export interface Customer {
  id: string
  name: string
  email: string
  projects: RelayConnection<Project>
  crm: CrmSystemBridge
}

export type SeatStatus = 'JOINED' | 'INVITED_PROJECT' | 'INVITED_CUSTOMER'

export interface Seat {
  name: string | null
  email: string
  isOwner: boolean
  status: SeatStatus
}

export interface CrmSystemBridge {
  id: string
  information: CrmCustomerInformation
  onboardingStatus: CrmOnboardingStatus
  customer: Customer
}

export interface CrmCustomerInformation {
  id: string
  name: string
  email: string
  isBeta: boolean
}

export type Environment = 'Node' | 'Browser'
export type GraphQLClient = 'fetch' | 'lokka' | 'relay' | 'apollo'

export type Example = 'ReactRelay' | 'ReactNativeApollo' | 'ReactApollo' | 'AngularApollo' | 'VueApollo'

export interface CrmOnboardingStatus {
  id: string
  gettingStarted: Step
  gettingStartedSkipped: boolean
  gettingStartedExample: Example | null
}

export interface Project {
  id: string
  name: string
  alias: string
  version: number
  models: RelayConnection<Model>
  relations: RelayConnection<Relation>
  actions: RelayConnection<Action>
  permanentAuthTokens: RelayConnection<PermanentAuthToken>
  authProviders: RelayConnection<AuthProvider>
  integrations: RelayConnection<Integration>
  actionSchema: string
  schema: string
  typeSchema: string
  enumSchema: string
  seats: RelayConnection<Seat>
  projectBillingInformation: ProjectBillingInformation
  region?: string
  packageDefinitions: RelayConnection<PackageDefinition>
  functions: RelayConnection<ServerlessFunction>
}

export interface ServerlessFunction {
  id: string
  name: string
  type?: FunctionType
  binding?: FunctionBinding
  webhookUrl: string
  _webhookUrl?: string
  webhookHeaders: string
  _webhookHeaders?: {[key: string]: string}
  inlineCode?: string
  auth0Id?: string
  logs: RelayConnection<Log>
  stats?: FunctionStats
  modelId?: string
  model?: Model
  operation?: RequestPipelineMutationOperation
  isActive: boolean
  query?: string
}

export interface FunctionStats {
  requestHistogram: number[]
  requestCount: number
  errorCount: number
  lastRequest: Date
}

export type RequestPipelineMutationOperation = 'CREATE' | 'UPDATE' | 'DELETE'

export interface Log {
  id: string
  requestId: string
  duration: number
  status: LogStatus
  timestamp: Date
  message: string
}

export type LogStatus = 'SUCCESS' | 'FAILURE'

export interface RequestPipelineMutationFunction extends ServerlessFunction {
  model: Model
  binding: FunctionBinding
}

export type FunctionType = 'WEBHOOK' | 'AUTH0'

export type FunctionBinding = 'TRANSFORM_ARGUMENT' | 'PRE_WRITE' | 'TRANSFORM_PAYLOAD'

export interface PackageDefinition {
  id: string
  definition: string
  email: string
  name: string
}

export interface ProjectBillingInformation {
  plan: string
}

export interface Integration {
  id: string
  isEnabled: boolean
  name: IntegrationNameType
  type: IntegrationTypeType
}

export interface SearchProviderAlgolia extends Integration {
  applicationId: string
  apiKey: string
  algoliaSyncQueries: RelayConnection<AlgoliaSyncQuery>
  algoliaSchema: string
}

export interface AlgoliaSyncQuery {
  id: string
  indexName: string
  fragment: string
  isEnabled: boolean
  model: Model
}

export type IntegrationNameType =
    'AUTH_PROVIDER_AUTH0'
  | 'AUTH_PROVIDER_DIGITS'
  | 'AUTH_PROVIDER_EMAIL'
  | 'SEARCH_PROVIDER_ALGOLIA'

export type IntegrationTypeType = 'AUTH_PROVIDER' | 'SEARCH_PROVIDER'

export type FieldType =
    'Relation'
  | 'Int'
  | 'String'
  | 'Boolean'
  | 'Enum'
  | 'Float'
  | 'DateTime'
  | 'Password'
  | 'Json'
  | 'GraphQLID'

export interface Field {
  id: string
  name: string
  description: string
  isRequired: boolean
  isList: boolean
  isUnique: boolean
  enumId?: string
  enum?: Enum
  isSystem?: boolean
  isReadonly?: boolean
  typeIdentifier?: FieldType
  defaultValue?: TypedValue
  enumValues: string[]
  reverseRelationField?: Field
  relatedModel?: Model
  relation?: Relation
  model?: Model
  migrationValue?: TypedValue
  constraints?: Constraint[]
  interface?: Interface
}

interface Interface {
  isSystem: boolean
  name: string
}

export interface Relation {
  id: string
  name: string
  isRequired: boolean
  description?: string
  leftModel: Model
  rightModel: Model
  fieldOnLeftModel: Field
  fieldOnRightModel: Field
  permissionSchema: string
  permissions: RelayConnection<RelationPermission>
  permissionQueryArguments: PermissionQueryArgument[]
}

export type UserType = 'EVERYONE' | 'AUTHENTICATED'

export type PermissionRuleType = 'NONE' | 'GRAPH' | 'WEBHOOK'

export interface Enum {
  id: string
  name: string
  values: string[]
}

export interface Model {
  id: string
  name: string
  namePlural: string
  fields: RelayConnection<Field>
  unconnectedReverseRelationFieldsFrom?: Field[]
  itemCount: number
  description: string
  isSystem: boolean
  permissions: RelayConnection<ModelPermission>
  permissionSchema: string
  permissionQueryArguments: PermissionQueryArgument[]
}

export interface PermissionQueryArgument {
  name: string
  typeName: string
  group: string
}

export interface ModelPermission {
  id: string
  fieldIds?: string[]
  ruleWebhookUrl?: string
  rule: Rule
  ruleName?: string
  ruleGraphQuery?: string
  applyToWholeModel: boolean
  isActive: boolean
  operation: Operation
  userType: UserType
}

export interface RelationPermission {
  id: string
  connect: boolean
  disconnect: boolean
  ruleWebhookUrl?: string
  rule: Rule
  ruleName?: string
  ruleGraphQuery?: string
  isActive: boolean
  userType: UserType
}

export type Rule = 'NONE' | 'GRAPH' | 'WEBHOOK'
export type Operation = 'READ' | 'CREATE' | 'UPDATE' | 'DELETE'

export type ActionTriggerType = 'MUTATION_MODEL' | 'MUTATION_RELATION'
export type ActionHandlerType = 'WEBHOOK'

export interface Action {
  id: string
  isActive: boolean
  description: string
  triggerType: ActionTriggerType
  handlerType: ActionHandlerType
  triggerMutationModel?: ActionTriggerMutationModel
  triggerMutationRelation?: ActionTriggerMutationRelation
  handlerWebhook?: ActionHandlerWebhook
}

export type ActionTriggerMutationModelMutationType = 'CREATE' | 'UPDATE' | 'DELETE'

export interface ActionTriggerMutationModel {
  id: string
  fragment: string
  model: Model
  mutationType: ActionTriggerMutationModelMutationType
}

export type ActionTriggerMutationRelationMutationType = 'ADD' | 'REMOVE'

export interface ActionTriggerMutationRelation {
  id: string
  fragment: string
  relation: Relation
  mutationType: ActionTriggerMutationRelationMutationType
}

export interface ActionHandlerWebhook {
  id: string
  url: string
}

export interface User {
  id: string
  createdAt?: string
}

export interface Node {
  id: string
  [key: string]: any
}

export interface PermanentAuthToken {
  id: string
  name: string
  token: string
}

export interface AuthProvider {
  id: string
  type: AuthProviderType
  isEnabled: boolean
  digits: AuthProviderDigits | null
  auth0: AuthProviderAuth0 | null
}

export type AuthProviderType = 'AUTH_PROVIDER_EMAIL' | 'AUTH_PROVIDER_DIGITS'
  | 'AUTH_PROVIDER_AUTH0' | 'anonymous-auth-provider'

export interface AuthProviderAuth0 {
  domain: string
  clientId: string
  clientSecret: string
}

export interface AuthProviderDigits {
  consumerKey: string
  consumerSecret: string
}

export interface OrderBy {
  fieldName: string
  order: 'ASC' | 'DESC'
}

export type FieldWidths = { [key: string]: number }

declare global {
  interface Element {
    scrollIntoViewIfNeeded(centerIfNeeded: boolean): void
  }
}

export interface TetherStep {
  step: Step
  title: string
  description?: string
  buttonText?: string
  copyText?: string
}

export type RelationPopupDisplayState = 'DEFINE_RELATION' | 'SET_MUTATIONS'
export type Cardinality = 'ONE_TO_ONE' | 'ONE_TO_MANY' | 'MANY_TO_ONE' | 'MANY_TO_MANY'

export type BreakingChangeIndicatorStyle = 'TOP' | 'RIGHT'

export type ConstraintType = 'REGEX' | 'CONTAINS' | 'STARTS_WITH' | 'ENDS_WITH' | 'EQUALS' | 'LENGTH'
export type Operator = 'EQ' | 'GT' | 'LT' | 'GTE' | 'LTE'

export interface Constraint {
  type: ConstraintType
  value: string
  lengthOperator?: Operator
}

export type PricingPlan = '2017-02-free' | '2017-02-project' | '2017-02-startup' | '2017-02-growth' | 'enterprise'

export interface PricingPlanInfo {
  name: string
  price: number
  maxStorage: number
  maxRequests: number
  maxSeats: number
  pricePerAdditionalMB: number
  pricePerThousandAdditionalRequests: number
}

export type CreditCardInputDisplayState = 'CREDIT_CARD_DATA' | 'ADDRESS_DATA'

export type Region = 'EU_WEST_1' | 'AP_NORTHEAST_1' | 'US_WEST_2'

export interface PermissionVariable {
  name: string
  typeIdentifier: FieldType
  category?: string
  isRequired: boolean
  isList: boolean
}

export interface Invoice {
  overageRequests: number
  overageStorage: number
  timestamp: string
  total: number
  usageRequests: number[]
  usageStorage: number[]
}

export interface CreditCardInfo {
  last4: string
  expMonth: number
  expYear: number
  cardHolderName: string
  addressLine1: string
  addressLine2?: string
  addressCity: string
  addressZip: string
  addressState: string
  addressCountry: string
}

export interface ProjectBillingInformation {
  plan: string
  invoices: Invoice[]
}
