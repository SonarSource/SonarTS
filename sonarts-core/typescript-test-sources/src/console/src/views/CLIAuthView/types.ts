export type AuthTrigger = 'auth' | 'init' | 'quickstart'
