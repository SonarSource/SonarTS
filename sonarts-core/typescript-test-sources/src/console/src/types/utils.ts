export type NotificationLevel = 'success' | 'error' | 'warning' | 'info'

export interface Notification {
  message: string
  level: NotificationLevel
}
export type ShowNotificationCallback = (notification: Notification) => void

export interface Query {
  query: string
  variables: string
  date: Date
}

export interface NonScalarValue {
  id: string
  [key: string]: TypedValue
}

export type ScalarValue = number | string | Date | boolean
export type AtomicValue = ScalarValue | NonScalarValue
export type TypedValue = ScalarValue | NonScalarValue | Array<AtomicValue>
