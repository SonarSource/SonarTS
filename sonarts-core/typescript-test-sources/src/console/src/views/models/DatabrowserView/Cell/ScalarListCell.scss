@import '~styles/variables';

.root {
  width: 700px;
  height: 100%;
  background: #fff;
}

.header {
  // height: 125px;
  background: #27AE60;
  // padding: 30px 30px 0;
}

.filter {
  display: flex;
  align-items: center;
  padding: 30px;
  font-size: 35px;
  color: #fff;
  overflow: hidden;

  input {
    padding: 0 0 0 30px;
    color: inherit;
  }

  svg {
    fill: #fff
  }

  .setNull {
    cursor: pointer;
    cursor: hand;

    &:hover {
      svg {
        fill: #D0021B;
      }
    }
  }
}
