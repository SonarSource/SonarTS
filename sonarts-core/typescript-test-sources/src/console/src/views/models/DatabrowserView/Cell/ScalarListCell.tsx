import * as React from 'react'
import PopupWrapper from '../../../../components/PopupWrapper/PopupWrapper'
import {getScalarEditCell, CellRequirements} from './cellgenerator'
import {AtomicValue} from '../../../../types/utils'
import {atomicValueToString, stringToValue} from '../../../../utils/valueparser'
import {Field} from '../../../../types/types'
const classes: any = require('./ScalarListCell.scss')
import * as Modal from 'react-modal'
import styled from 'styled-components'
import {variables, $p, Icon} from 'graphcool-styles'
import * as cx from 'classnames'
import {isEqual} from 'lodash'
import modalStyle from '../../../../utils/modalStyle'

interface State {
  values: AtomicValue[]
  filter: string
  newValue: string
  editingIndex: number
}

const Button = styled.button`
      padding: ${variables.size16};
      font-size: ${variables.size16};
      border: none;
      background: none;
      color: ${variables.gray50};
      border-radius: 2px;
      cursor: pointer;
      transition: color ${variables.duration} linear;

      &:hover {
        color: ${variables.gray70};
      }
    `

const CancelButton = styled(Button)`
      padding: 0;
      background: transparent;
      color: ${variables.gray50};

      &:hover {
        color: ${variables.gray60};
      }
    `

const SaveButton = styled(Button)`
      background: ${variables.green};
      color: ${variables.white};

      &:hover {
        color: ${variables.white80};
      }
    `

const InputWrapper = styled.div`
  input {
    padding: 0 !important;
  }
`

const DateInputWrapper = styled(InputWrapper)`
  > div, > div > div, > div > div > div {
    position: static !important;
  }
`

const NewInputWrapper = styled(InputWrapper)`
  input {
    color: ${variables.blue} !important;
  }
`

const NewDateInputWrapper = styled(NewInputWrapper)`
  > div, > div > div, > div > div > div {
    position: static !important;
  }
`

const Item = styled.div`
  i {
    opacity: 0;
    transition: .3s opacity;
  }
  
  &:hover {
    i {
      opacity: 1;
    }
  }
`

const customModalStyle = {
  overlay: {
    ...modalStyle.overlay,
    background: 'rgba(255,255,255,.4)',
  },
  content: {
    ...modalStyle.content,
    background: 'white',
    width: 400,
  },
}

export default class ScalarListCell extends React.Component<CellRequirements, State> {

  private atomicField: Field = Object.assign({}, this.props.field, {isList: false})

  constructor(props: CellRequirements) {
    super(props)
    this.state = {
      filter: '',
      newValue: null,
      values: this.props.value ? this.props.value.slice() : [],
      editingIndex: null,
    }
  }

  getEditRequirements(index: number) {
    return {
      value: this.state.values[index],
      field: this.atomicField,
      projectId: this.props.projectId,
      nodeId: this.props.nodeId,
      methods: {
        save: this.handleValueChange.bind(this, index),
        cancel: () => null,
        onKeyDown: this.onKeyDown,
      },
    }
  }

  render() {
    const newRequirements: CellRequirements = {
      value: this.state.newValue,
      field: this.atomicField,
      projectId: this.props.projectId,
      nodeId: this.props.nodeId,
      inList: true,
      methods: {
        save: this.handleValueChange.bind(this, -1),
        cancel: () => null,
        onKeyDown: this.onKeyDown,
      },
    }
    return (
      <Modal
        isOpen
        style={customModalStyle}
        contentLabel='Edit scalar list'
        onRequestClose={this.handleOutsideClick}
      >
        <style jsx global>{`
          .scalar-list-cell {
            .add-icon {
              @p: .bgBlue, .br100, .pointer;
            }
          }

          .scalar-list-cell .remove-icon {
            @p: .bgrRed, .br100, .flex, .itemsCenter, .justifyCenter, .pointer;
            padding: 8px;
            height: 30px;
            width: 30px;
          }
        `}</style>
        <style jsx>{`
          input {
            background: transparent;
          }
        `}</style>
        <div className='scalar-list-cell'>
          <div
            className={cx(
              $p.bgWhite,
              $p.flex,
              $p.flexColumn,
              $p.relative,
            )}
          >
            <div className={classes.header}>
              <div className={classes.filter}>
                <Icon
                  src={require('assets/new_icons/search.svg')}
                  width={30}
                  height={30}
                  color={variables.blue}
                />
                <input
                  type='text'
                  placeholder='Filter...'
                  value={this.state.filter}
                  onChange={(e: any) => null}
                />
                <Icon
                  src={require('assets/icons/delete.svg')}
                  color={variables.red}
                  width={30}
                  height={30}
                  onClick={() => {
                    this.props.methods.save(null)
                    this.props.methods.cancel()
                  }}
                />
              </div>
            </div>
            <div
              className={cx(
                $p.flexAuto,
              )}
            >
              <div
                className={cx(
                  $p.f14,
                  $p.flex,
                  $p.flexRow,
                  $p.justifyBetween,
                  $p.black50,
                  $p.bb,
                  $p.bBlack10,
                  $p.pa25,
                  $p.itemsCenter,
                )}
                onClick={ () => this.setState({editingIndex: -1} as State)}
              >
                {this.state.editingIndex === -1 && (
                  this.props.field.typeIdentifier === 'DateTime' ? (
                    <NewDateInputWrapper className={$p.flexAuto}>
                      {getScalarEditCell(newRequirements)}
                    </NewDateInputWrapper>
                  ) : (
                    <NewInputWrapper className={$p.flexAuto}>
                      {getScalarEditCell(newRequirements)}
                    </NewInputWrapper>
                  )
                )}
                {!(this.state.editingIndex === -1) && (
                  <span className={classes.addNewText}>
                    {this.state.newValue ? (
                      <div>{atomicValueToString(this.state.newValue, this.props.field, true)}</div>
                    ) : (
                      'Add new item ...'
                    )}
                  </span>
                )}
                <Icon
                  className={'add-icon ' + (this.state.newValue ? '' : classes.disabled)}
                  onClick={() => this.addNewValue()}
                  src={require('graphcool-styles/icons/stroke/add.svg')}
                  stroke
                  strokeWidth={2.5}
                  width={30}
                  height={30}
                  color={variables.white}
                />
              </div>
              <div
                className={cx($p.overflowYScroll, $p.flexAuto)}
                style={{maxHeight: 'calc(100vh - 294px)'}}
              >
                {this.state.values.map((value, index) => (
                  <Item
                    key={index}
                    className={cx(
                      $p.f14,
                      $p.flex,
                      $p.flexRow,
                      $p.justifyBetween,
                      $p.black50,
                      $p.bb,
                      $p.bBlack10,
                      $p.pa25,
                      $p.itemsCenter,
                    )}
                    onClick={() => this.setEditingIndex(index)}
                  >
                    {this.state.editingIndex === index ? (
                      this.props.field.typeIdentifier === 'DateTime' ? (
                        <DateInputWrapper className={$p.flexAuto}>
                          {getScalarEditCell(this.getEditRequirements(index))}
                        </DateInputWrapper>
                      ) : (
                        <InputWrapper className={$p.flexAuto}>
                          {getScalarEditCell(this.getEditRequirements(index))}
                        </InputWrapper>
                      )
                    ) : (
                      <div>{atomicValueToString(value, this.props.field, true)}</div>
                    )}
                    <Icon
                      className='remove-icon'
                      src={require('graphcool-styles/icons/stroke/cross.svg')}
                      stroke
                      strokeWidth={4}
                      width={30}
                      height={30}
                      onClick={() => this.handleDeleteValue(index)}
                      color={variables.white}
                    />
                  </Item>
                ))}
              </div>
            </div>
            <div className={cx(
              $p.bgBlack04,
              $p.pa25,
              $p.flex,
              $p.flexRow,
              $p.justifyBetween,
              $p.itemsCenter,
            )}>
              <CancelButton onClick={this.props.methods.cancel}>
                Cancel
              </CancelButton>
              <SaveButton onClick={this.handleClose}>
                Save
              </SaveButton>
            </div>
          </div>
        </div>
      </Modal>
    )
  }

  private setEditingIndex = (editingIndex: number) => {
    this.setState({editingIndex} as State)
  }

  private onKeyDown = (e: any) => {
    if (e.keyCode === 13) {
      this.addNewValue(stringToValue(e.target.value, this.props.field, true) as AtomicValue)
      e.stopPropagation()
      return
    }
    this.props.methods.onKeyDown(e)
  }

  private handleOutsideClick = () => {
    if (isEqual(this.state.values, this.props.value)) {
      this.props.methods.cancel()
      return
    }

    // only ask, if there was change
    graphcoolConfirm('This action could lead to data loss.', 'Unsaved Changes')
      .then(() => {
        this.props.methods.cancel()
      })
  }

  private handleClose = () => {
    let values = this.state.values
    if (this.state.values.length === 0 && (
      typeof this.state.newValue !== 'undefined' && this.state.newValue !== null)
    ) {
      values = [this.state.newValue]
    }
    this.props.methods.save(values, false)
  }

  private handleDeleteValue = (index: number) => {
    const result = this.state.values.slice(0)
    result.splice(index, 1)
    this.setState({
      values: result,
    } as State)
  }

  private handleValueChange = (index: number, value: AtomicValue) => {
    if (index === -1) {
      this.setState({newValue: value} as State)
    } else {
      const values = this.state.values
      values[index] = value
      this.setState({values} as State)
    }
  }

  private addNewValue = (value: AtomicValue = null) => {
    // we need this complicated check, as the boolean field false should be allowed
    if ((this.state.newValue === null || this.state.newValue === undefined)
        && (value === null || value === undefined)) {
      return
    }

    const values = this.state.values.slice(0)
    const newValue = value || this.state.newValue

    values.unshift(newValue)
    this.setState({
      newValue: '',
      values,
    } as State)
  }
}
