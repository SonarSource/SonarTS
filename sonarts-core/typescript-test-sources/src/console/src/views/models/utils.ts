import {isScalar} from '../../utils/graphql'
import {Field, FieldWidths} from '../../types/types'
import {TypedValue, NonScalarValue, ScalarValue} from '../../types/utils'
import {stringToValue, valueToString, getFieldTypeName} from '../../utils/valueparser'
import calculateSize from 'calculate-size'
import * as Immutable from 'immutable'
import {isNonScalarList} from '../../utils/graphql'

export function emptyDefault(field: Field): TypedValue {

  if (field.isRequired) {
    return null
  }

  const value = function (): any {
    switch (field.typeIdentifier) {
      case 'Int':
        return 0
      case 'Float':
        return 0
      case 'DateTime':
        return new Date().toISOString()
      case 'String':
        return ''
      case 'Boolean':
        return false
      case 'Enum':
        return field.enumValues.length > 0 ? field.enumValues[0] : null
      default:
        return null
    }
  }()

  if (!field.isList) {
    return value
  }

  return []
}

function valueToGQL(value: TypedValue, field: Field): string {
  if (value === null && !field.isRequired) {
    return 'null'
  }

  if (!isScalar(field.typeIdentifier)) {
    if (field.isList) {
      // we can safely assume that it's a list at this point
      const values = value as any[]
      if (values.length === 0) {
        return '[]'
      } else {
        return `[${values.map(v => `"${v.id}"`).join(',')}]`
      }
    }
    return `"${(value as NonScalarValue).id}"`
  }
  if (field.typeIdentifier === 'Enum') {
    if (field.isList) {
      return `[${(value as ScalarValue[]).join(',')}]`
    }
    return value as string
  }

  return JSON.stringify(value)
}

export function toGQL(value: TypedValue, field: Field): string {
  let key = isScalar(field.typeIdentifier) ? field.name : field.isList ? `${field.name}Ids` : `${field.name}Id`

  if (value === null && field.isRequired) {
    return ''
  }

  const output = valueToGQL(value, field)
  return `${key}: ${output}`
}

export function compareFields(a: Field, b: Field): number {
  if (a.name === 'id') {
    return -1
  }
  if (b.name === 'id') {
    return 1
  }
  return a.name.localeCompare(b.name)
}

export function getFirstInputFieldIndex(fields: Field[]): number {
  let inputIndex
  const hasInputField = fields.some((field, index) => {
    if (isNonScalarList(field) || field.name === 'id' || field.isReadonly) {
      return false
    } else {
      inputIndex = index
      return true
    }
  })

  if (hasInputField) {
    return inputIndex
  } else {
    return null
  }
}

export function getDefaultFieldValues(fields: Field[]): { [key: string]: any } {
  return fields.filter((f) => f.name !== 'id')
    .mapToObject(
      (field) => field.name,
      (field) => ({
        value: stringToValue(field.defaultValue, field) || emptyDefault(field),
        field: field,
      }),
    )
}

export function calculateFieldColumnWidths (width: number,
                                            fields: Field[],
                                            nodes: any,
                                           ): FieldWidths {
  const cellFontOptions = {
    font: 'Open Sans',
    fontSize: '12px',
  }

  const headerFontOptions = {
    font: 'Open Sans',
    fontSize: '12px',
  }

  const widths = fields.mapToObject(
    (field) => field.name,
    (field) => {
      switch (field.name) {
        case 'id':
          return 220
        case 'createdAt':
        case 'updatedAt':
          return 200
      }

      let cellWidths = nodes
      .filter(node => !!node)
      .map(node => node.hasOwnProperty('get') ? node.get(field.name) : node[field.name])
      .map(value => valueToString(value, field, false))
      .map(str => calculateSize(str, cellFontOptions).width + 41)
      if (cellWidths.hasOwnProperty('toArray')) {
        cellWidths = cellWidths.toArray()
      }

      const headerWidth = calculateSize(`${field.name} ${getFieldTypeName(field)}`, headerFontOptions).width + 90

      const maxWidth = Math.max(...cellWidths, headerWidth)
      const lowerLimit = field.typeIdentifier === 'DateTime' ? 250 : 150
      const upperLimit = 400

      const idExtra = field.name === 'id' ? 50 : 0
      const relationExtra = field.typeIdentifier === 'Relation' ? 50 : 0

      return (
        maxWidth > upperLimit ?
        upperLimit :
        (maxWidth < lowerLimit ? lowerLimit : maxWidth)
      ) + idExtra + relationExtra
    },
  )

  const totalWidth = fields.reduce((sum, {name}) => sum + widths[name], 0)
  const fieldWidth = width - 34 - 250
  if (totalWidth < fieldWidth) {
    fields.forEach(({name}) => {
      if (!['id', 'createdAt', 'updatedAt'].includes(name)) {
        widths[name] = (widths[name] / totalWidth) * fieldWidth
      }
    })
  }
  return widths
}
