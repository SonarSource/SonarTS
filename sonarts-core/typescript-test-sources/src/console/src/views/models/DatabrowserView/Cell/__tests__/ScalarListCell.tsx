import * as React from 'react' // tslint:disable-line
import ScalarListCell from '../ScalarListCell'
import { shallow } from 'enzyme'
import { shallowToJson } from 'enzyme-to-json'
import {TypedValue} from '../../../../../types/utils'
import {FieldType} from '../../../../../types/types'

test('ScalarListCell renders', () => {

  const save = jest.fn((value: TypedValue) => {  /* */ })
  const cancel = jest.fn()
  const onKeyDown = jest.fn()
  const field = {
    id: 'cip3p48sj001e1jsmghwkdt2k',
    name: 'description',
    description: '',
    isReadonly: true,
    isList: true,
    isSystem: true,
    typeIdentifier: 'Int' as FieldType,
    relatedModel: null,
    isUnique: true,
    isRequired: false,
    enumValues: [],
    model: null,
  }

  const component = shallow(
    <ScalarListCell
      value={[1,2,3]}
      nodeId={null}
      projectId={null}
      methods={{
        save, cancel, onKeyDown,
      }}
      field={field}
    />,
  )

  expect(shallowToJson(component)).toMatchSnapshot()

})
