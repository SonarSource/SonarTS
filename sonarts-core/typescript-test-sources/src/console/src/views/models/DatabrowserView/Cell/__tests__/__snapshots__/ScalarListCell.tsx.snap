exports[`test ScalarListCell renders 1`] = `
<Modal
  ariaHideApp={true}
  closeTimeoutMS={0}
  contentLabel="Edit scalar list"
  isOpen={true}
  onRequestClose={[Function]}
  parentSelector={[Function]}
  portalClassName="ReactModalPortal"
  shouldCloseOnOverlayClick={true}
  style={
    Object {
      "content": Object {
        "background": "white",
        "border": "none",
        "borderRadius": 2,
        "bottom": "initial",
        "boxShadow": "0 1px 7px rgba(0,0,0,.2)",
        "height": "auto",
        "left": "initial",
        "overflow": "visible",
        "padding": 0,
        "position": "relative",
        "right": "initial",
        "top": "initial",
        "width": 400,
      },
      "overlay": Object {
        "alignItems": "center",
        "background": "rgba(255,255,255,.4)",
        "backgroundColor": "rgba(255,255,255,.9)",
        "display": "flex",
        "justifyContent": "center",
        "zIndex": 100,
      },
    }
  }>
  <style
    global={true}
    jsx={true}>
    
              .scalar-list-cell {
                .add-icon {
                  @p: .bgBlue, .br100, .pointer;
                }
              }
    
              .scalar-list-cell .remove-icon {
                @p: .bgrRed, .br100, .flex, .itemsCenter, .justifyCenter, .pointer;
                padding: 8px;
                height: 30px;
                width: 30px;
              }
            
  </style>
  <style
    jsx={true}>
    
              input {
                background: transparent;
              }
            
  </style>
  <div
    className="scalar-list-cell">
    <div
      className="bgWhite flex flexColumn relative">
      <div
        className="header">
        <div
          className="filter">
          <Icon
            color="rgba(42,126,210,1)"
            height={30}
            src="test-file-stub"
            width={30} />
          <input
            onChange={[Function]}
            placeholder="Filter..."
            type="text"
            value="" />
          <Icon
            color="rgba(242,92,84,1)"
            height={30}
            onClick={[Function]}
            src="test-file-stub"
            width={30} />
        </div>
      </div>
      <div
        className="flexAuto">
        <div
          className="f14 flex flexRow justifyBetween black50 bb bBlack10 pa25 itemsCenter"
          onClick={[Function]}>
          <span
            className="addNewText">
            Add new item ...
          </span>
          <Icon
            className="add-icon disabled"
            color="#fff"
            height={30}
            onClick={[Function]}
            src="test-file-stub"
            stroke={true}
            strokeWidth={2.5}
            width={30} />
        </div>
        <div
          className="overflowYScroll flexAuto"
          style={
            Object {
              "maxHeight": "calc(100vh - 294px)",
            }
          }>
          <styled.div
            className="f14 flex flexRow justifyBetween black50 bb bBlack10 pa25 itemsCenter"
            onClick={[Function]}>
            <div>
              1
            </div>
            <Icon
              className="remove-icon"
              color="#fff"
              height={30}
              onClick={[Function]}
              src="test-file-stub"
              stroke={true}
              strokeWidth={4}
              width={30} />
          </styled.div>
          <styled.div
            className="f14 flex flexRow justifyBetween black50 bb bBlack10 pa25 itemsCenter"
            onClick={[Function]}>
            <div>
              2
            </div>
            <Icon
              className="remove-icon"
              color="#fff"
              height={30}
              onClick={[Function]}
              src="test-file-stub"
              stroke={true}
              strokeWidth={4}
              width={30} />
          </styled.div>
          <styled.div
            className="f14 flex flexRow justifyBetween black50 bb bBlack10 pa25 itemsCenter"
            onClick={[Function]}>
            <div>
              3
            </div>
            <Icon
              className="remove-icon"
              color="#fff"
              height={30}
              onClick={[Function]}
              src="test-file-stub"
              stroke={true}
              strokeWidth={4}
              width={30} />
          </styled.div>
        </div>
      </div>
      <div
        className="bgBlack04 pa25 flex flexRow justifyBetween itemsCenter">
        <styled.button
          onClick={[Function]}>
          Cancel
        </styled.button>
        <styled.button
          onClick={[Function]}>
          Save
        </styled.button>
      </div>
    </div>
  </div>
</Modal>
`;
